src/specification/inverse.ts

```diff
@@ -290,9 +398,69 @@ function simplifyMatch(match: Match, given: string): Match | null {
     };
 }
 
-function expectsSuccessor(condition: Condition, given: string) {
-    return condition.type === "path" &&
-        condition.labelRight === given &&
-        condition.rolesRight.length === 0 &&
-        condition.rolesLeft.length > 0;
+function expectsSuccessor(condition: Condition, given: string) {
+    return condition.type === "path" &&
+        condition.labelRight === given &&
+        condition.rolesRight.length === 0 &&
+        condition.rolesLeft.length > 0;
+}
+
+/**
+ * Validates that a specification meets the required invariant properties:
+ * 1. All matches except the first must have at least one path condition
+ * 2. Path conditions must reference labels that appear earlier in the specification
+ * 
+ * @param specification The specification to validate
+ * @returns true if the specification is valid
+ * @throws Error with descriptive message if the specification violates invariants
+ */
+export function validateSpecificationInvariant(specification: Specification): boolean {
+    // Collect all available labels in order: givens first, then matches in order
+    const availableLabels = new Set<string>();
+    
+    // Add all given labels
+    for (const given of specification.given) {
+        availableLabels.add(given.name);
+    }
+    
+    // Process each match in order
+    for (let i = 0; i < specification.matches.length; i++) {
+        const match = specification.matches[i];
+        
+        // First match doesn't need path conditions
```

Comment:
Once a sequence of matches is converted back to a specification, the first match -- the one without path conditions -- is promoted to a given. Therefore, this invariant check should not exclude the first match of a specification.